cuda/private/actions/compile.bzl

```diff
@@ -34,14 +37,27 @@ def compile(
     cuda_feature_config = cuda_helper.configure_features(ctx, cuda_toolchain, requested_features = [ACTION_NAMES.cuda_compile])
     artifact_category_name = cuda_helper.get_artifact_category_from_action(ACTION_NAMES.cuda_compile, pic, rdc)
 
-    ret = []
+    basename_counter = {}
+    src_and_indexed_basenames = []
     for src in srcs:
         # this also filter out all header files
         basename = cuda_helper.get_basename_without_ext(src.basename, ALLOW_CUDA_SRCS, fail_if_not_match = False)
         if not basename:
             continue
+        basename_index = basename_counter.setdefault(basename, default=0)
+        basename_counter[basename] += 1
+        src_and_indexed_basenames.append((src, basename, basename_index))
 
+    ret = []
+    for src, basename, basename_index in src_and_indexed_basenames:
+        filename = None
         filename = cuda_helper.get_artifact_name(cuda_toolchain, artifact_category_name, basename)
+        # Objects are placed in _objs/<tgt_name>/<filename>.
+        # For files with the same basename, say srcs = ["kernel.cu", "foo/kernel.cu", "bar/kernel.cu"], we get
+        # _objs/<tgt_name>/0/kernel.<ext>, _objs/<tgt_name>/1/kernel.<ext>, _objs/<tgt_name>/2/kernel.<ext>.
+        # Otherwise, the index is not presented.
+        if basename_counter[basename] > 1:
+            filename = "{}/{}".format(basename_index, filename)
```

Comment:
Could we instead go to `kernel.0.<ext>`, `kernel.1.<ext>` instead to avoid introducing the extra inode usage on the filesystem that creating extra directories incurs?

Reply:
There will possible be
```
foo/kernel.cu -> kernel.0.cu ???
bar/kernel.cu -> kernel.1.cu
kernel.0.cu -> kernel.0.cu ???
```

Then the problem will be even annoying. 🤣

Reply:
cc rules also has the same behavior.